core/src/client/mod.rs

```diff
@@ -172,21 +183,26 @@ pub trait TransportReceiverT: 'static {
 	async fn receive(&mut self) -> Result<ReceivedMessage, Self::Error>;
 }
 
+/// Convert the given values to a [`crate::params::ArrayParamsBuilder`], or empty tuple as expected by a
+/// jsonrpsee Client (http or websocket).
+///
+/// # Panics
+///
+/// Panics if the serialization of parameters fails.
 #[macro_export]
-/// Convert the given values to a [`jsonrpsee_types::ParamsSer`] as expected by a jsonrpsee Client (http or websocket).
 macro_rules! rpc_params {
+	() => {
+		$crate::client::__reexports::EmptyParams
+	};
 	($($param:expr),*) => {
 		{
-			let mut __params = vec![];
+			let mut __params = $crate::client::__reexports::ArrayParamsBuilder::new();
 			$(
-				__params.push($crate::client::__reexports::to_json_value($param).expect("json serialization is infallible; qed."));
+				__params.insert($param).expect(format!("Parameter `{}` cannot be serialized", stringify!($param)).as_str());
```

Comment:
Nit: we need to allocate a string to pass to `expect` even if we hope it's never called.

I wonder whether the compiler is clever here. But maybe safer to avoid the alloc unless we actually hit an error with something like:

```rust
if let Err(e) = __params.insert($param) {
    panic!("Parameter `{}` cannot be serialized: {}", stringify!($param), e)
}
```

Also, no need to use `__params` really (no chance of name conflicts); `params` would be fine :)

Reply:
(also  also; you know this already but we can remove the "empty" case and return an empty ArrayParamsBuilder.build() too so that the macro is consistently the same type and such (it is afterall very much an "array params" builder, in the same way `vec![]` always builds `Vec`s)

Reply:
Yep that is the case: https://rust.godbolt.org/z/E33fsWW13, have added the `if let Err` pattern to avoid an allocation. Thanks!